### PR TITLE
[Data] Improve notebook widget display

### LINKED ISSUE #31225: [Datasets] `Dataset` representations don't work out-of-the-box on notebooks
### What happened + What you expected to happen

**On Google Colab**:

At first, you get warning telling you to upgrade the `ipywidgets` version.

image

If you do that and then restart the runtime you get a new problem:

> Support for third party widgets (widgets outside of the ipywidgets package) needs to be enabled separately. Support for these widgets will be loaded from a CDN external from Colab.

image

**On Anyscale**:

Instead of a readable output, you get HTML like `VBox(children=(HTML(value='<h2>Dataset</h2>'), Tab(children=(HTML(value='<div class="scrollableTable jp-Render…`:

image

### Versions / Dependencies

Ray: master

### Reproduction script

See above.

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by peytondmurray**
## Why are these changes needed?

This PR aims to fix some outstanding issues with notebook ipywidget display.

1. Changes error messaging for the `ipywidgets` soft dependency to include explicit instructions to 

* Install/upgrade `ipywigets` as appropriate
* Restart the notebook (i.e. jupyter) server, without which widgets will not be properly displayed.

I've also added a number of tests to ensure these decorators are working correctly.

2. Switch from using `_ipython_display_` to `_repr_mimebundle_` for displaying reprs of `DataParallelTrainer` and `Datastream` objects.

This change is motivated by the fact that when using `_ipython_display_` to display widgets, it is the responsibility of the author of the `_ipython_display_` method to identify the right repr to display depending on the display capabilities of the frontend. This introduces additional complexity, since `IPython.get_ipython()` is usually the way by which people detect whether they are running in a notebook, but the result of this function depends on the kernel being used and doesn't directly tell you about the display capabilities of the frontend.

Instead, a better way to do this is to provide a variety of reprs (e.g. an ipywidget, a simple text repr, etc...) and let the frontend decide which one to display. This is what the `_repr_mimebundle_` function is meant to do. Here's what this looks like in practice:

image

Here, `Datastream._repr_mimebundle_()` returns both a text repr and an ipywidget repr. If you run this in the IPython REPL, you get the text repr, but if you run it in a notebook you get the ipywidget. Nice! :tada: 

**Note**: This PR also makes `ipywidgets>=8` a hard dependency of `ray[data]`, `ray[tune]`, and `ray[train]`, at the request of @amogkam. As a side effect of this, the requirements for `myst-nb` and `myst-parser` have been updated to their latest versions.

## Related issue number

Closes #31225. This is not a foolproof solution for this display issue since it still requires the user to restart the server, but now the instructions that show up about installing `ipywigets` and restarting the server are now explicit about what to do to fix it.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by scottjlee — APPROVED** [review]

**Comment by peytondmurray** [comment]
One test was broken when `IPython` was not installed. Fixed, retrying now.

**Comment by peytondmurray** [comment]
Okay, I checked this on Google Colab and the fix seems to be working:

image

**Comment by peytondmurray** [comment]
Arrgh, missed a debugging statement from before. Going to write an RFC to add a check for this this to the pre-commit hooks, but for now it should be fixed.

**Comment by peytondmurray** [comment]
@amogkam This is currently breaking minimal install tests, because in one of the tests there's an `import train`, which depends on `ray[air]` which depends on `ray[data]`, which depends on `ipywidgets`, which isn't included in the minimal install. Here's the import chain:

```
(ActorWithLibImport pid=71436)   File "/home/pdmurray/Desktop/workspace/ray/python/ray/tests/test_usage_stats.py", line 1434, in __init__
(ActorWithLibImport pid=71436)     from ray import train  # noqa: F401
(ActorWithLibImport pid=71436)   File "/home/pdmurray/Desktop/workspace/ray/python/ray/train/__init__.py", line 2, in <module>
(ActorWithLibImport pid=71436)     from ray.train.backend import BackendConfig
(ActorWithLibImport pid=71436)   File "/home/pdmurray/Desktop/workspace/ray/python/ray/train/backend.py", line 4, in <module>
FAILED(ActorWithLibImport pid=71436)     from ray.air.checkpoint import Checkpoint
(ActorWithLibImport pid=71436)   File "/home/pdmurray/Desktop/workspace/ray/python/ray/air/__init__.py", line 2, in <module>
(ActorWithLibImport pid=71436)     from ray.air.config import (
(ActorWithLibImport pid=71436)   File "/home/pdmurray/Desktop/workspace/ray/python/ray/air/config.py", line 24, in <module>
(ActorWithLibImport pid=71436)     from ray.data.preprocessor import Preprocessor
(ActorWithLibImport pid=71436)   File "/home/pdmurray/Desktop/workspace/ray/python/ray/data/__init__.py", line 8, in <module>
(ActorWithLibImport pid=71436)     from ray.data.dataset import Dataset
(ActorWithLibImport pid=71436)   File "/home/pdmurray/Desktop/workspace/ray/python/ray/data/dataset.py", line 1, in <module>
(ActorWithLibImport pid=71436)     from ray.data.datastream import Datastream
(ActorWithLibImport pid=71436)   File "/home/pdmurray/Desktop/workspace/ray/python/ray/data/datastream.py", line 25, in <module>
(ActorWithLibImport pid=71436)     import ipywidgets
(ActorWithLibImport pid=71436) ModuleNotFoundError: No module named 'ipywidgets'
```

I don't know how you want to solve this - we could make `ipywidgets` a hard dependency of `ray`, or...modify the test? I'm not sure, but I'll defer to others on this. @scottjlee FYI

**Comment by richardliaw** [comment]
Hmm, doesn't ray[data] have other dependencies as well? What happens to those?

**Comment by peytondmurray** [comment]
@richardliaw Yes it does, but none of them cause the issue that `ipywidgets` did because:

* `numpy`: Not listed as a direct dependency of the other ray packages, but is a dependency of a bunch of other stuff so probably gets installed anyway. Looking more closely at `ci/env/install-minimal.sh`, there's a line which also installs `pytest` and `numpy` in that script, though I don't think it's explicitly listed anywhere. To add to the confusion, the numpy version required here for `ray[data]` is different than the one used by ray itself. In summary, I can say for sure numpy gets installed and thus won't cause the issue that `ipywidgets` did above, but it's hard to say what version gets installed.
* `pandas`: Also a dependency of tune, train, and air
* `pyarrow`: Also a dependency of tune, train, and air
* `fsspec`: Not a dependency of tune or train, but this is listed in `python/requirements.txt`, which is installed when `ci/env/install-minimal.sh` is run on the CI runner.

@amogkam Given the complexity of the CI, I actually don't see where the ray extras are being installed. Do you know where the install command is for these? Searching for `ray[data]` etc inside `ci/` doesn't return anything, and looking through all the `pip install <whatever>` commands I never was able to find a place where `pip install -v .[data]` or similar was being run. Do you have any additional information about how the CI job is run here?

**Review by ericl — CHANGES_REQUESTED** [review]
I don't think we can add ipywidgets as a hard dep. It should be a soft dependency, where we have a placeholder message saying "please pip install ipywidgets" to get an enhanced output.

**Comment by amogkam** [comment]
Can you elaborate more on the concerns?

It's a very small dependency and improves UX considerably. To me, it's a no-brainer to add as a dependency for libraries. Note that it is not added as a dependency for `pip install ray`.

**Comment by amogkam** [comment]
ok reverted back, can you review again @ericl

File: python/ray/widgets/util.py [hunk 1]
```diff
@@ -102,6 +102,43 @@ def wrapped(*args, **kwargs):
...
+    """
+    text = (
+        "Run `pip install {}ipywidgets`, then restart "
+        "the notebook server for rich notebook output."
+    )
+
+    if in_notebook():
+        return ensure_notebook_deps(
+            ["ipywidgets", version],
+            missing_message=text.format(""),
```
Comment (ericl) [inline python/ray/widgets/util.py:128]: Could we still print the string repr if the notebook deps aren't found? It seems like the current code will fail to print any repr if the deps are unavailable.

**Review by ericl — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]
approve for requirements docker.txt

File: python/ray/widgets/util.py (hunk 1 above)
Comment (amogkam) [reply python/ray/widgets/util.py:128]: updated

**Comment by ericl** [comment]
ERROR: Cannot install myst-parser==0.15.2 and myst-parser==0.18.1 because these package versions have conflicting dependencies.
  |  
  | The conflict is caused by:
  | The user requested myst-parser==0.18.1
  | The user requested myst-parser==0.15.2
 

<br class="Apple-interchange-newline">

**Comment by peytondmurray** [comment]
Awesome, thanks for the assist moving back to a soft dependency!

**Pull request merged**